cli/lsp/language_server.rs

```diff
@@ -1299,7 +1299,24 @@ impl Inner {
     );
 
     let response = if !all_actions.is_empty() {
-      Some(all_actions)
+      if self.config.client_capabilities.code_action_disabled_support {
+        Some(all_actions)
+      } else {
+        // if the client does not support disabled code actions, then we should
+        // drop them
+        Some(
+          all_actions
+            .into_iter()
+            .filter(|ca| {
+              if let CodeActionOrCommand::CodeAction(ca) = ca {
+                ca.disabled.is_none()
+              } else {
+                true
+              }
+            })
+            .collect(),
+        )
```

Comment:
Seeing that `response` is `None` when `all_actions.is_empty() == true`, and that the filter here could result in an empty list, it should probably evaluate to `None` in that case?

I'd try to make it less branch-y, e.g.:
```rust
let code_action_disabled_support =
  self.config.client_capabilities.code_action_disabled_support;

let actions = all_actions
  .into_iter()
  .filter(|ca| {
    code_action_disabled_support
      || matches!(ca, CodeActionOrCommand::CodeAction(ca) if ca.disabled.is_none())
  })
  .collect();

let response = if actions.is_empty() {
  None
} else {
  Some(actions)
};

// or if you want to shave off a few lines:
let response = actions.is_empty().then(|| actions);
```
There won't be more than a handful of actions. Filtering should be essentially no cost.

Reply:
Thanks for the feedback.  Makes sense!